src/main/java/org/openRealmOfStars/player/ship/Ship.java

```diff
@@ -300,21 +309,48 @@ public void regenerateShield() {
    * @return true if has energy
    */
   public boolean hasComponentEnergy(final int index) {
+      return hasComponent()
+              && (hasRemainingEnergy(index)
+                      || !isComponentRequireEnergy(index));
+  }
+
+/**
+ * @return true if has component
+ */
+private boolean hasComponent() {
+    return components.size() > 0;
+}
+
+/**
+ * @param index Component index
+ * @return true if component require energy
+ */
+private boolean isComponentRequireEnergy(final int index) {
+    return components.get(index).getEnergyRequirement() > 0;
+}
+
+/**
+ * @param index Component index
+ * @return true if remaining energy >= 0
+ */
+private boolean hasRemainingEnergy(final int index) {
+    return getRemainingEnergy(index) >= 0;
+}
+
+/**
+ * @param index Component index
+ * @return remaining energy
+ */
+private int getRemainingEnergy(final int index) {
     int energy = getTotalEnergy();
-    for (int i = 0; i < components.size(); i++) {
+    for (int i = 0; i <= index; i++) {
```

Comment:
I think it would be better check that index < components.size is true before entering for loop.